### PR TITLE
Adding line integral visualization to vector fields

### PR DESCRIPTION
**Pull request by Adnan-Ali-Ahmad**

### CODE REVIEW

File: src/osyris/plot/render.py [hunk 1]
```diff
@@ -30,13 +30,24 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
         "vec": "quiver",
         "vector": "quiver",
         "stream": "streamplot",
+        "lic": "line_integral_convolution",
         None: config.parameters["render_mode"],
         "image": "pcolormesh",
         "imshow": "pcolormesh"
     }
 
+    for item in data: #check if user requested a line integral conv
```
Comment (nvaytet) [inline src/osyris/plot/render.py:39]: I think this mechanism is a little strange.

To be more in-line with how the other layer modes work, I think the `lic` mode should just behave like a vec mode, where the values for the colormap would be an additional argument, e.g. `color`.

Basically, you would make a `lic` plot by doing
```Py
osyris2.plane(
            {"data": data["hydro"]["velocity"], "mode": "lic", "color": data["hydro"]["density"], "norm": "log"},
            dx=dx * osyris2.units("au"),
            origin=center,
            direction="z",
            resolution=512)
```

I think this means that this code block would go away, and the `line_integral_convolution` wrapper would call the `pcolormesh` wrapper if the `color` is specified, and then fiddle with the colors by adding the `lic` component on top.

(Note that using the `color` argument is how you would set colors of the arrows on a quiver mode.)

File: src/osyris/plot/render.py [hunk 2]
```diff
@@ -46,7 +57,7 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
             del item["params"]["cbar"]
         else:
             cbar = True
-
+
```
Comment (nvaytet) [inline src/osyris/plot/render.py:60]: trailing spaces

File: src/osyris/plot/render.py [hunk 3]
```diff
@@ -30,13 +30,24 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
...
+        try:
+            if item["mode"] == "lic":
+                data.append(data.pop(data.index(item))) #move lic plot to end of data list
+                item["params"]["alpha"] = .3 #add alpha param to make the lic blend with the main image
+        except KeyError:
+            pass
+
     mpl_objects = []
-    for item in data:
+    for item in data:
```
Comment (nvaytet) [inline src/osyris/plot/render.py:50]: trailing spaces

File: src/osyris/plot/wrappers.py [hunk 4]
```diff
@@ -6,6 +6,8 @@
 from .. import config
 from ..core import Array
 from pint.quantity import Quantity
+import lic as lic
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:9]: The import of `lic` should be moved inside the `line_integral_convolution` function to avoid making `lic` a hard dependency for `osyris`

File: src/osyris/plot/wrappers.py [hunk 5]
```diff
@@ -40,12 +42,13 @@ def pcolormesh(ax, x, y, z, **kwargs):
...
     default_args = {
-        "shading": "nearest",
+        "origin": "lower",
+        "extent": [np.min(x),np.max(x),np.min(y),np.max(y)]
     }
     default_args.update(kwargs)
     if "cmap" not in kwargs:
         kwargs["cmap"] = config.parameters["cmap"]
-    return ax.pcolormesh(x, y, z, **default_args)
+    return ax.imshow(z,**default_args)
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:51]: I would like to keep `pcolormesh` because it allows me to do things like setting a log-scale on the x or y axes, which is more tricky with `imshow`.

File: src/osyris/plot/wrappers.py [hunk 6]
```diff
@@ -126,3 +129,22 @@ def scatter(ax, x, y, data, **kwargs):
...
+    lic_data_clip_rescale = (lic_data_clip-lim[0])/(lim[1]-lim[0])
+    lic_data_rgba[...,3] = lic_data_clip_rescale * 1
+    
+    args = [lic_data_rgba]
+    plot_args = {**kwargs}
+    plot_args["cmap"] = "binary"
+    plot_args["extent"] = [np.min(x),np.max(x),np.min(y),np.max(y)]
+    plot_args["origin"] = "lower"
+    
+    return ax.imshow(*args, **plot_args)
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:150]: I was able to play with the colors of `pcolormesh`, but in a non-standard way.
Setting the `alpha` to something else than 1 on a `pcolormesh` is not great because some edge artifacts start to appear (see e.g. https://stackoverflow.com/questions/52100747/changing-facecolor-alpha-value-gives-unwanted-edges-matplotlib-pcolormesh)
So we want to play with the facecolors directly.

I can do something like
```Py
p = pcolormesh(ax=ax,
               x=x,
               y=y,
               z=z[..., 2],
               **kwargs)
# pcolormesh needs a draw call before the array of facecolors is defined
ax.get_figure().canvas.draw()
facecolors = p.get_facecolors()
flat_shape = facecolors.shape[0]
newcolors = facecolors.reshape(*z[..., 2].shape, 4)
newcolors[..., :3] *= 0.6 * (lic_res.reshape(*lic_res.shape, 1) - 0.5) + 1.0
p.set_facecolors(np.clip(newcolors.reshape(flat_shape, 4), 0, 1))
return p
```
but it doesn't look quite as good as your picture, maybe you know best how to fiddle with the contrast, or maybe it's not possible this way and you have to use `
Screenshot at 2021-11-02 18-36-18
imshow` and alpha values?

**Review by nvaytet — CHANGES_REQUESTED** [review]
Thanks for the contribution, I think it will be a great addition to the plotting capabilities.

Some suggestions as how to make this more in line with the rest of the code/api

File: src/osyris/plot/wrappers.py (hunk 6 above)
Comment (nvaytet) [reply src/osyris/plot/wrappers.py:150]: On second thought, it might be just simpler to call the pcolormesh wrapper if there is a color requested, and then overlay an image with imshow and alpha blending?

Comment (Adnan-Ali-Ahmad) [reply src/osyris/plot/wrappers.py:150]: Try this: `newcolors[..., :3] *= .6 + (lic_data_rgba[...,:3] - 0.5) + 1.0` , with lic_data_rgba as computed in the line_integral_convolution wrapper function. You'll get better results due to the better contrast inside the lic, however I don't really think this is a solution because the results will depend quite heavily on the colormap the user requests, which means we'll need to fiddle the values in this line of code to get optimal-looking results for each colormap. Also, should you use any non-sequential colormaps (such as jet for example), you'll get some pretty bad looking plots since we're applying a linear transformation on all the pixel color values.

To circumvent this problem, we need to work with the opacity of the pixels (fourth axis in newcolors and lic_data_rgba), but unfortunately this does seem to cause pixel-edge artifacts as you mentioned. So I agree more with your second thought, it might be easier to simply use imshow with alpha blending should the user request a lic, and pcolormesh for all other purposes.

File: src/osyris/plot/render.py (hunk 1 above)
Comment (Adnan-Ali-Ahmad) [reply src/osyris/plot/render.py:39]: You're right, that makes more sense. I'll try implementing it.

File: src/osyris/plot/render.py [hunk 7]
```diff
@@ -51,15 +52,24 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
...
             need_cbar = True
         if (func == "scatter") and ("c" in item["params"]):
             if not isinstance(item["params"]["c"], str):
                 need_cbar = True
         if need_cbar and cbar:
-            cb = plt.colorbar(mpl_objects[-1], ax=ax, cax=None)
-            cb.set_label(make_label(name=item["name"], unit=item["unit"]))
+            if func == "line_integral_convolution":
+                cb = plt.colorbar(mpl_objects[-1].images[0], ax=ax, cax=None)
+                cb.set_label(make_label(name=item["params"]["color"].name, unit=item["params"]["color"].unit))
```
Comment (Adnan-Ali-Ahmad) [inline src/osyris/plot/render.py:69]: Small issue with this line of code here. It seems item["params"]["color"].unit is displaying the correct units but with a number in front of them (ex. 1.0 g / cm**3) and I'm not entirely sure how to give this a clean fix.

File: src/osyris/plot/wrappers.py [hunk 8]
```diff
@@ -126,3 +126,50 @@ def scatter(ax, x, y, data, **kwargs):
...
+    from matplotlib.cm import ScalarMappable
+    from matplotlib.colors import LogNorm
+
+    #compute line integral convolution
+    try:
+        lic_res = lic.lic(z[...,1], z[...,0], length=kwargs["length"])
+        del kwargs["length"]
+    except KeyError:
+        lic_res = lic.lic(z[...,1], z[...,0], length=30)
+
```
Comment (Adnan-Ali-Ahmad) [inline src/osyris/plot/wrappers.py:145]: "length" is now also an additional plotting parameter for the lic. Its default value is 30 which gives good results for most plots, but the user might want a different value at times.

**Review by Adnan-Ali-Ahmad — COMMENTED** [review]
The LIC now behaves as a mode, and the result blends the image with the "color" requested by the user. Let me know if there's anything else that needs changing

File: src/osyris/plot/parser.py [hunk 9]
```diff
@@ -18,7 +18,6 @@ def parse_layer(layer,
                 vmax=None,
                 operation=None,
                 **kwargs):
-
```
Comment (nvaytet) [inline src/osyris/plot/parser.py:21]: Unnecessary change?

File: src/osyris/plot/plane.py [hunk 10]
```diff
@@ -77,6 +77,7 @@ def plane(*layers,
     to_render = []
     operations = []
     to_scatter = []
+
```
Comment (nvaytet) [inline src/osyris/plot/plane.py:80]: Unnecessary change?

Comment (nvaytet) [inline src/osyris/plot/plane.py:80]: Unnecessary change?

File: src/osyris/plot/render.py [hunk 11]
```diff
@@ -51,15 +52,23 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
             getattr(wrappers, func)(ax, x, y, item["data"], **item["params"]))
 
         need_cbar = False
+
+        if func == "line_integral_convolution":
+            if "color" in item["params"]:
+                need_cbar = True
+
```
Comment (nvaytet) [inline src/osyris/plot/render.py:59]:
```suggestion
        if func == "line_integral_convolution" and "color" in item["params"]:
            need_cbar = True
```

File: src/osyris/plot/render.py [hunk 12]
```diff
@@ -51,15 +52,23 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
...
         if func in ["contourf", "pcolormesh"]:
             need_cbar = True
         if (func == "scatter") and ("c" in item["params"]):
             if not isinstance(item["params"]["c"], str):
                 need_cbar = True
         if need_cbar and cbar:
-            cb = plt.colorbar(mpl_objects[-1], ax=ax, cax=None)
-            cb.set_label(make_label(name=item["name"], unit=item["unit"]))
+            if func == "line_integral_convolution":
```
Comment (nvaytet) [inline src/osyris/plot/render.py:66]: I would try to avoid the if statement here by making the `line_integral_convolution` return an image object, instead of the axes.
I can see above that `mpl_objects` uses `.append`, so it is actually accepting only a single object.
If you want to return both images, you could modify `mpl_objects` to use `.extend` instead. You would then have to modify the other wrapper functions to return a list, containing a single object.
But this is probably a useful change for the future anyway, as other wrappers may generate more than one mpl objects.

File: src/osyris/plot/render.py [hunk 13]
```diff
@@ -51,15 +52,23 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
...
+                cb = plt.colorbar(mpl_objects[-1].images[0], ax=ax, cax=None)
+                cb.set_label(make_label(name=item["params"]["color"].name, unit=item["params"]["color"].unit.units))
+            else:
+                cb = plt.colorbar(mpl_objects[-1], ax=ax, cax=None)
+                cb.set_label(make_label(name=item["name"], unit=item["unit"]))
             cb.ax.yaxis.set_label_coords(-1.1, 0.5)
-
     out["objects"] = mpl_objects
-    return out
+    return out
```
Comment (nvaytet) [inline src/osyris/plot/render.py:74]: Missing new line at the end of file

File: src/osyris/plot/wrappers.py [hunk 14]
```diff
@@ -126,3 +126,50 @@ def scatter(ax, x, y, data, **kwargs):
...
+    plot_args = {**kwargs}
+    plot_args["extent"] = [np.min(x),np.max(x),np.min(y),np.max(y)]
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:147]: Note that here the `x` and `y` are pixel centers, and we actually need the pixel edges for the true extent.

File: src/osyris/plot/wrappers.py (hunk 8 above)
Comment (nvaytet) [reply src/osyris/plot/wrappers.py:145]: I would use a different approach here. I would explicitly list `length` in the arguments (and `color` as well for that matter), and simply give it a default value of 30. Then you only need a single line here.
Look at the `labels` argument in the `contour` wrapper for an example.

File: src/osyris/plot/wrappers.py [hunk 15]
```diff
@@ -126,3 +126,50 @@ def scatter(ax, x, y, data, **kwargs):
...
+    plot_args["origin"] = "lower"
+
+    alpha_blending = True
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:150]: If you list `color` in the function args, you should be able to replace `alpha_blending` with `color is not None`

File: src/osyris/plot/wrappers.py [hunk 16]
```diff
@@ -126,3 +126,50 @@ def scatter(ax, x, y, data, **kwargs):
...
+    try:
+        del plot_args["color"]
+    except KeyError:
+        alpha_blending = False
+
+    if alpha_blending:
+        plot_args_color = {**plot_args}
+        plot_args_color["norm"] = LogNorm()
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:158]: Why is the norm always logarithmic?

File: src/osyris/plot/wrappers.py [hunk 17]
```diff
@@ -126,3 +126,50 @@ def scatter(ax, x, y, data, **kwargs):
...
+        ax.imshow(z[...,2], **plot_args_color)
+
+        plot_args["alpha"] = .3
+
+    #amplify contrast on lic
+    lim=(.2,.5)
+    lic_data_clip = np.clip(lic_res,lim[0],lim[1])
+    lic_data_rgba = ScalarMappable(norm=None, cmap="binary").to_rgba(lic_data_clip)
+    lic_data_clip_rescale = (lic_data_clip-lim[0])/(lim[1]-lim[0])
+    lic_data_rgba[...,3] = lic_data_clip_rescale * 1
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:169]: Why is the `* 1` needed here?

Comment (nvaytet) [inline src/osyris/plot/wrappers.py:169]: Why is the `* 1` needed here?

File: src/osyris/plot/wrappers.py [hunk 18]
```diff
@@ -126,3 +126,50 @@ def scatter(ax, x, y, data, **kwargs):
...
+
+    #plot the lic
+    plot_args["cmap"] = "binary"
+    ax.imshow(lic_data_rgba, **plot_args)
+
+    return ax
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:175]: Missing new line

**Review by nvaytet — CHANGES_REQUESTED** [review]
Some more comments.

You should also add an example in the `plotting` notebook in the docs.

**Comment by nvaytet** [comment]
Note that you'll have to run the code source through `yapf` and `flake8` before you can get the builds to pass.
Let me know if you need help on how to do that.

File: src/osyris/plot/wrappers.py (hunk 16 above)
Comment (Adnan-Ali-Ahmad) [reply src/osyris/plot/wrappers.py:158]: My mistake, I'll fix this.

File: src/osyris/plot/wrappers.py (hunk 17 above)
Comment (Adnan-Ali-Ahmad) [reply src/osyris/plot/wrappers.py:169]: It's not, I was fiddling with some values and not changing lic_data_clip_rescale was fine as is.

File: src/osyris/plot/render.py (hunk 12 above)
Comment (Adnan-Ali-Ahmad) [reply src/osyris/plot/render.py:66]: I can avoid the extra if statement and still render the correct colorbar the following way:

`if "cbar" in item["params"]:
            cbar = item["params"]["cbar"]
            del item["params"]["cbar"]
        else:
            cbar = True

        mpl_objects.extend(
            getattr(wrappers, func)(ax, x, y, item["data"], **item["params"]))

        need_cbar = False

        ind_render = -1
        name = item["name"]
        unit = item["unit"]

        if func == "line_integral_convolution" and "color" in item["params"]:
            need_cbar = True
            ind_render = -2
            name = item["params"]["color"].name
            unit = item["params"]["color"].unit.units

        if func in ["contourf", "pcolormesh"]:
            need_cbar = True
        if (func == "scatter") and ("c" in item["params"]):
            if not isinstance(item["params"]["c"], str):
                need_cbar = True
        if need_cbar and cbar:
            cb = plt.colorbar(mpl_objects[ind_render], ax=ax, cax=None)
            cb.set_label(make_label(name=name, unit=unit))
            cb.ax.yaxis.set_label_coords(-1.1, 0.5)`

Since the line_integral_convolution wrapper returns two images, we need to instruct the renderer to only create a colorbar for the first ax.imshow that gets returned (which is the first item in the return list of said wrapper function, hence the ind_render = -2). Let me know if you're okay with this syntax.

File: src/osyris/plot/wrappers.py (hunk 14 above)
Comment (Adnan-Ali-Ahmad) [reply src/osyris/plot/wrappers.py:147]: The to_bin_edges function in tools.py doesn't work; it's trying to concatenate arrays with simple floats. I think this is an adequate fix: 
```
def to_bin_edges(x):
    """
    Convert array centers to edges
    """
    centers = to_bin_centers(x)
    dx = x[1]-x[0]
    return np.append(centers-dx,np.array([centers[-1],centers[-1]+dx]))
```

**Comment by Adnan-Ali-Ahmad** [comment]
Also, I noticed a bug; if the user requests a "vec" mode before a density plot, the density plot overwrites the quivers in the image. This can be fixed by rendering the quivers after all other modes have been rendered:

(in render.py)
``` 
for item in data:
        try:
            if item["mode"] == "vec":
                data.append(data.pop(data.index(item))) #move item to end of data list
        except KeyError:
            pass
```

This will make sure that
```
plane = osyris.plane({"data": data["hydro"]["velocity"], "mode": "vec"},
			{"data": data["hydro"]["density"], "norm": "log","cmap":"viridis"},
             dx=dx * osyris.units("au"),
             origin=center,
             direction="z",plot=True, resolution = 1000)
```
still renders the quiver plot.

**Review by Adnan-Ali-Ahmad — COMMENTED** [review]
added most requested changes (and some more). also, some newspaces and import orders have been changed to conform to flake8 formatting.

**Comment by nvaytet** [comment]
> Also, I noticed a bug; if the user requests a "vec" mode before a density plot, the density plot overwrites the quivers in the image. This can be fixed by rendering the quivers after all other modes have been rendered:
> 
> (in render.py)
> 
> ```
> for item in data:
>         try:
>             if item["mode"] == "vec":
>                 data.append(data.pop(data.index(item))) #move item to end of data list
>         except KeyError:
>             pass
> ```
> 
> This will make sure that
> 
> ```
> plane = osyris.plane({"data": data["hydro"]["velocity"], "mode": "vec"},
> 			{"data": data["hydro"]["density"], "norm": "log","cmap":"viridis"},
>              dx=dx * osyris.units("au"),
>              origin=center,
>              direction="z",plot=True, resolution = 1000)
> ```
> 
> still renders the quiver plot.

Nice catch.
However, for this one, I think the mechanism of changing the order of the layers sounds a bit brittle.
I think a better way would be to set the `zorder` of the contour, quiver and streamlines to something higher than the images.
I will fix this in a separate PR, no need to make changes for that here.

**Comment by Adnan-Ali-Ahmad** [comment]
Actually forget about the last change I made. It's early in the morning and I shouldn't be making any commits.

File: src/osyris/plot/render.py [hunk 19]
```diff
@@ -47,19 +49,30 @@ def render(x=None, y=None, data=None, logx=False, logy=False, ax=None):
...
+        mpl_objects.extend(
             getattr(wrappers, func)(ax, x, y, item["data"], **item["params"]))
 
         need_cbar = False
+
+        ind_render = -1
+        if func != "scatter":
+            name = item["name"]
+            unit = item["unit"]
+
```
Comment (Adnan-Ali-Ahmad) [inline src/osyris/plot/render.py:61]: documentation build should work now

**Comment by Adnan-Ali-Ahmad** [comment]
I'm not very familiar with test suites, do we have to add lic as a requirement in order for the build to succeed?

**Comment by nvaytet** [comment]
> I'm not very familiar with test suites, do we have to add lic as a requirement in order for the build to succeed?

Ah yes, you have to add `lic` to the `docs/requirements.txt` file so that it installs it when trying to build the docs :-)

**Comment by Adnan-Ali-Ahmad** [comment]
done :)

File: src/osyris/core/tools.py [hunk 20]
```diff
@@ -38,9 +39,8 @@ def to_bin_edges(x):
     Convert array centers to edges
     """
     centers = to_bin_centers(x)
-    left = centers[0] - (x[1] - x[0])
-    right = centers[-1] + (x[-1] - x[-2])
-    return np.concatenate(np.concatenate(left, centers), right)
+    dx = x[1] - x[0]
+    return np.append(centers - dx, np.array([centers[-1], centers[-1] + dx]))
```
Comment (nvaytet) [inline src/osyris/core/tools.py:43]: While I can see that there is an issue with my implementation (it actually does not work, it throws an error), I do not agree with this change.
Your implementation assumes that all the cells have the same width, which is not necessarily the case.

I suggest to replace with
```Py
def to_bin_edges(x):
    centers = to_bin_centers(x)
    left = centers[0] - (x[1] - x[0])
    right = centers[-1] + (x[-1] - x[-2])
    return np.append(np.insert(centers, 0, left), right)
```

Consider
```Py
a = np.arange(11.)
to_bin_edges(a)
```
Both your implementation and the suggested one above would return
```
[-0.5  0.5  1.5  2.5  3.5  4.5  5.5  6.5  7.5  8.5  9.5 10.5]
```
as expected. However, for
```Py
b = np.arange(11.)**2
to_bin_edges(b)
```
your implementation returns
```
[-0.5  1.5  5.5 11.5 19.5 29.5 41.5 55.5 71.5 89.5 90.5 91.5]
```
while what I am suggesting yields
```
[ -0.5   0.5   2.5   6.5  12.5  20.5  30.5  42.5  56.5  72.5  90.5 109.5]
```
In the figure below, you can see the `edges1` (my suggestion) in orange are a better representation of bin edges
Screenshot at 2021-11-09 23-14-29

File: src/osyris/plot/wrappers.py [hunk 21]
```diff
@@ -123,6 +124,51 @@ def scatter(ax, x, y, data, **kwargs):
...
+    lic_res = lic.lic(z[..., 1], z[..., 0], length=length)
+
+    plot_args = {**kwargs}
+    plot_args["extent"] = [
+        np.min(xedges), np.max(xedges),
+        np.min(yedges), np.max(yedges)
+    ]
+    plot_args["origin"] = "lower"
+
+    axes = []
```
Comment (nvaytet) [inline src/osyris/plot/wrappers.py:153]:
```suggestion
    images = []
```

**Review by nvaytet — CHANGES_REQUESTED** [review]

File: src/osyris/core/tools.py (hunk 20 above)
Comment (Adnan-Ali-Ahmad) [reply src/osyris/core/tools.py]: I hadn't thought of that. Nice.

**Review by Adnan-Ali-Ahmad — COMMENTED** [review]
Added the requested changes.

**Review by nvaytet — APPROVED** [review]
Nice work. Thanks again for the contribution!
